bin/clover/src/pipeline-steps/addDefaultPropsAndSockets.ts

```diff
@@ -0,0 +1,149 @@
+import { PkgSpec } from "../bindings/PkgSpec.ts";
+import _ from "lodash";
+import {
+  createObjectProp,
+  createScalarProp,
+  isExpandedPropSpec,
+} from "../spec/props.ts";
+import { getSiFuncId } from "../spec/siFuncs.ts";
+import { attrFuncInputSpecFromSocket, createSocket } from "../spec/sockets.ts";
+
+export function addDefaultPropsAndSockets(specs: PkgSpec[]): PkgSpec[] {
+  const newSpecs = [] as PkgSpec[];
+
+  for (const spec of specs) {
+    const schema = spec.schemas[0];
+
+    if (!schema) {
+      console.log(
+        `Could not generate default props and sockets for ${spec.name}: missing schema`,
+      );
+      continue;
+    }
+    const schemaVariant = schema.variants[0];
+
+    if (!schemaVariant) {
+      console.log(
+        `Could not generate default props and sockets for ${spec.name}: missing variant`,
+      );
+      continue;
+    }
+
+    const domain = schemaVariant.domain;
+    if (!isExpandedPropSpec(domain)) {
+      console.log(
+        `Could not generate default props and sockets for ${spec.name}: domain has no metadata`,
+      );
+      continue;
+    }
+    if (domain.kind !== "object") {
+      console.log(
+        `Could not generate default props and sockets for ${spec.name}: domain is not object`,
+      );
+      continue;
+    }
+
+    // Extra prop
+    const extraProp = createObjectProp("extra", domain.metadata.propPath);
+
+    // Create PropUsageMap
+    {
+      const propUsageMapProp = createScalarProp(
+        "PropUsageMap",
+        "string",
+        extraProp.metadata.propPath,
+      );
+      const propUsageMap = {
+        createOnly: [] as string[],
+        updatable: [] as string[],
+      };
+
+      const queue = _.cloneDeep(domain.entries);
+
+      while (queue.length > 0) {
+        const prop = queue.pop();
+        if (!prop || !isExpandedPropSpec(prop)) break;
+
+        if (prop.metadata.createOnly) {
+          propUsageMap.createOnly.push(prop.name);
+        } else if (!prop.metadata.readOnly) {
+          propUsageMap.updatable.push(prop.name);
+        }
+
+        if (prop.kind === "object") {
+          prop.entries.forEach((p) => queue.unshift(p));
+        }
+      }
+
+      propUsageMapProp.data.defaultValue = JSON.stringify(propUsageMap);
```

Comment:
I wonder how this handles on updates. Like, if we generate a new asset with a new createOnly prop, will it update the value?

Reply:
It will skip the value when creating the payload